File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> map = new HashMap<>();
        
        map.put('}','{');
        map.put(')', '(');
        map.put(']','[');
        
        for(int i = 0; i<s.length(); i++){
            if(map.containsKey(s.charAt(i))){
                
                if(stack.isEmpty() || stack.pop()!= map.get(s.charAt(i))){
                    return false;
                }
            } else {
                stack.add(s.charAt(i));
            }
        }
        
        return stack.empty();

        
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        
        stack = []
        inverse_dict = {'(': ')', '{':'}', '[':']'}
        
        for i in range(len(s)):
            if len(stack) == 0:
                stack.append(inverse_dict.get(s[i]))
            
            else:
                if stack[-1] == s[i]:
                    stack.pop()
                else:
                    stack.append(inverse_dict.get(s[i]))
        
        return len(stack) == 0